R/settings.R

```diff
@@ -77,6 +78,97 @@ read_config_file <- function(config_file) {
   config
 }
 
+validate_config_file <- function(config, config_file, defaults) {
+  matched <- names(config) %in% names(defaults)
+  if (!all(matched)) {
+    warning("Found unused settings in config '", config_file, "': ", toString(names(config)[!matched]))
+  }
+
+  validate_regex(config,
+    c("exclude", "exclude_next", "exclude_start", "exclude_end", "exclude_linter", "exclude_linter_sep")
+  )
+  validate_character_string(config, c("encoding", "cache_directory", "comment_token"))
+  validate_true_false(config, c("comment_bot", "error_on_lint"))
+  validate_linters(config$linters)
+  validate_exclusions(config$exclusions)
+}
+
+is_character_string <- function(x) is.character(x) && length(x) == 1L && !is.na(x)
+is_valid_regex <- function(str) !inherits(tryCatch(grepl(str, ""), condition = identity), "condition")
+is_single_regex <- function(x) is_character_string(x) && is_valid_regex(x)
+is_true_false <- function(x) is.logical(x) && length(x) == 1L && !is.na(x)
```

Comment:
Can we already use `isTRUE(x) || isFALSE(x)`?

Reply:
Even so we could backport, but note that it's redundant:

```
isTRUE <- function(x) is.logical(x) && length(x) == 1L && !is.na(x) && x
isFALSE <- function(x) is.logical(x) && length(x) == 1L && !is.na(x) && !x
```

This version avoids double-testing 3/4 conditions

Reply:
Huh? Isn't it `identical(x, TRUE)` and `identical(x FALSE)` respectively?

Reply:
(note the comment above about why they abandoned that implementation)